### PR TITLE
[AMORO-2861]: Support display Hudi table metadata in Amoro Dashboard

### LINKED ISSUE #2861: [Subtask][Feature]: Display Apache Hudi metadata in AMS Dashboard
### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/amoro/issues?q=is%3Aissue) and found no similar issues.

### Description

- Display Hudi tables in HadoopCatalog and HiveCatalog
- Show Schema and properties in Table Detail Page
- Show Hudi timeline and commit information in Table Snapshot Page
- Show Hudi compact and clean instant in Table Optimizing Page 
- Statistic Hudi parititions and files and display them on Table Files Pages

### Parent issue

#220 

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by baiyangtx**
## Why are the changes needed?
 

Close #2861.

## Brief change log
 
- Add hudi table catalog support 
- Implement hudi format catalog for Hadoop metastore.
- Implement hudi format catalog for Hive metastore

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by baiyangtx** [comment]
Support choose Hudi format when create hive/hadoop catalog.
image

Support display Hudi tables in catalog, with hudi icon.
image

Support display Hudi table details, include table summary, scheme, metrics, table type and properties
image

Support display partitions and files in partitions.
Due to limit of hudi api, some field without values.
image
image

Support display timeline as snapshots
image
image
Due to limit of hudi api, records statistic is missed, and file count is incorrect for deltacommit.

Support display compact/cluster instant as optimizing process
image
image
Only show completed instant, and show compact as minor optimize, cluster as major optimize.

File: amoro-ams/amoro-ams-server/pom.xml [hunk 1]
```diff
@@ -186,6 +186,13 @@
             <artifactId>derby</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+            <version>8.0.33</version>
```
Comment (zhoujinsong) [inline amoro-ams/amoro-ams-server/pom.xml:192]: How about declaring the version in the properties section in parent pom.xml?

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/dashboard/HudiTableDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+      List<PartitionFileBaseInfo> fileInPartition =
+          fsStream.flatMap(fs -> fileSliceToFileStream(partition, fs)).collect(Collectors.toList());
+      files.addAll(fileInPartition);
+    }
+    return files;
+  }
+
+  @Override
+  public List<DDLInfo> getTableOperations(AmoroTable<?> amoroTable) {
+    return Lists.newArrayList();
```
Comment (zhoujinsong) [inline amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/dashboard/HudiTableDescriptor.java:266]: Maybe add some comment to tell why?

File: amoro-ams/amoro-ams-server/pom.xml [hunk 3]
```diff
@@ -283,6 +290,22 @@
             <artifactId>flink-runtime-web</artifactId>
         </dependency>
 
+        <!--  apache hudi dependencies -->
+        <dependency>
```
Comment (zhoujinsong) [inline amoro-ams/amoro-ams-server/pom.xml:294]: Adding this template into dependencyManagement may simplify the configuration of sub modules.

File: amoro-core/src/main/java/org/apache/amoro/formats/hudi/HudiHadoopCatalog.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/** Hudi catalog implement in hadoop filesystem. */
+public class HudiHadoopCatalog implements FormatCatalog {
```
Comment (zhoujinsong) [inline amoro-core/src/main/java/org/apache/amoro/formats/hudi/HudiHadoopCatalog.java:46]: It seems Hudi does not provide its own catalog api. It shocks me.   0.0

File: amoro-core/src/test/java/org/apache/amoro/hive/HMSMockServer.java [hunk 5]
```diff
@@ -388,4 +388,11 @@ private void runScript(Connection connection, Reader reader) throws SQLException
       }
     }
   }
+
+  public static void main(String[] args) throws InterruptedException {
```
Comment (zhoujinsong) [inline amoro-core/src/test/java/org/apache/amoro/hive/HMSMockServer.java:392]: Seems to be some not needed codes here.

**Review by zhoujinsong — COMMENTED** [review]
Thanks for the great work.

I left some comments, PTAL.

**Comment by zhoujinsong** [comment]
Some check style errors exist, which you may want to fix. @baiyangtx

File: amoro-ams/amoro-ams-server/pom.xml (hunk 1 above)
Comment (zhoujinsong) [reply amoro-ams/amoro-ams-server/pom.xml:192]: And I am not sure why this dependency is necessary.

File: amoro-core/src/main/java/org/apache/amoro/formats/hudi/HudiHadoopCatalog.java (hunk 4 above)
Comment (baiyangtx) [reply amoro-core/src/main/java/org/apache/amoro/formats/hudi/HudiHadoopCatalog.java:46]: Yes，and the action load action are different in spark and flink...

File: amoro-ams/amoro-ams-server/pom.xml (hunk 1 above)
Comment (baiyangtx) [reply amoro-ams/amoro-ams-server/pom.xml:192]: The purpose of adding this dependency is to make it easier to start AMS in IDEA and debug directly.

File: amoro-ams/amoro-ams-server/pom.xml (hunk 3 above)
Comment (baiyangtx) [reply amoro-ams/amoro-ams-server/pom.xml:294]: This dependency has different configuration in ams-server and ams-core

**Comment by czy006** [comment]
cc @majin1102 @zhoujinsong

**Review by zhoujinsong — APPROVED** [review]
LGTM.

Thanks for the contribution!

**Comment by yhf20071** [comment]
when use getTableSnaphots, some exception occurs

**Comment by yhf20071** [comment]
when use snapshot detail ,the exception is 
Getting all partition paths with file system listing sequentially can be very slow. This should not be invoked.

**Comment by yhf20071** [comment]
when use snapshot detail ,the exception is 
Getting all partition paths with file system listing sequentially can be very slow. This should not be invoked.

**Comment by zhoujinsong** [comment]
@yhf20071 Thanks for the feedback.

Can you create GH issues to help the community track these bugs?

**Pull request merged**